File: rectangular/binarytree.go
package rectangular

import "math/rand"

func (m *Maze) ApplyBinaryTree() {
	m.name = "Binary Tree"
	for r := 0; r < m.rows; r++ {
		for c := 0; c < m.cols; c++ {
			// Pick between linking vertically or horizontally.
			rnd := rand.Intn(2)
			if r+1 == m.rows && c+1 == m.cols {
				// Neither will work.
				rnd = -1
			} else if r+1 == m.rows {
				// Linking rows won't work, so use columns.
				rnd = 1
			} else if c+1 == m.cols {
				// Linking columns won't work, so use rows.
				rnd = 0
			}
			switch rnd {
			case 0:
				// Link to next row.
				m.g.AddEdge(
					CellHash(Cell{row: r, col: c}),
					CellHash(Cell{row: r + 1, col: c}),
				)
			case 1:
				// Link to next column.
				m.g.AddEdge(
					CellHash(Cell{row: r, col: c}),
					CellHash(Cell{row: r, col: c + 1}),
				)
			}
		}
	}
}

File: c/algo/binary_tree.c
#include "binary_tree.h"

#include <stdlib.h>
#include <string.h>
#include <time.h>

void apply_binary_tree(Maze *maze) {
    srand(time(NULL));
    strcpy(maze->name, "Binary Tree");
    for (int x = 0; x < maze->width; x++) {
        for (int y = 0; y < maze->height; y++) {
            Cell c = {.x = x, .y = y};
            Cell north = north_of(c);
            Cell east = east_of(c);

            // Pick between linking vertically or horizontally.
            int rnd = rand() % 2;

            if (!maze_has_cell(maze, north) && !maze_has_cell(maze, east)) {
                // Neither will work.
                rnd = -1;
            } else if (!maze_has_cell(maze, north)) {
                // Linking rows won't work, so use columns.
                rnd = 1;
            } else if (!maze_has_cell(maze, east)) {
                // Linking columns won't work, so use rows.
                rnd = 0;
            }

            if (rnd == 0) {
                // Link to next row.
                maze_add_edge(maze, c, north);
            } else if (rnd == 1) {
                // Link to next column.
                maze_add_edge(maze, c, east);
            }
        }
    }
}
